### PR TITLE
Implement ftl::inspect_iterator

### PR DESCRIPTION
**Pull request by Blinkuu**

### CODE REVIEW

**Comment by Blinkuu** [comment]
This is still WIP

File: include/ftl/array.hpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ class array {
     std::fill_n(data_, N, value);
   }
 
-  constexpr auto swap(array &other) -> void {
+  constexpr auto swap(array &other) noexcept -> void {
```
Comment (BIGbadEL) [inline include/ftl/array.hpp]: Template parameters should be passed explicitly.

File: include/ftl/array.hpp [hunk 2]
```diff
@@ -208,6 +208,20 @@ struct ftl::from_iterator_trait<ftl::enumerate_iterator<Iter>, ftl::array<std::t
   }
 };
 
+template<typename Iter, typename Callable, typename Item, std::size_t N>
+struct ftl::from_iterator_trait<ftl::inspect_iterator<Iter, Callable>, ftl::array<Item, N>> {
+  [[nodiscard]] constexpr static auto from_iter(const inspect_iterator<Iter, Callable> &iter) -> array<Item, N> {
```
Comment (BIGbadEL) [inline include/ftl/array.hpp]: should be explicitly ftl::array.

File: include/ftl/inspect_iterator.hpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  inspect_iterator(Iter iterator, Callable callable) : iterator_{ std::move(iterator) }, callable_{ std::move(callable) } {
+  }
+
+private:
+  template<typename Collection>
+  [[nodiscard]] auto collect_impl() const -> Collection {
+    return from_iterator_trait<inspect_iterator<Iter, Callable>, Collection>::from_iter(*this);
+  }
+
+  [[nodiscard]] constexpr auto count_impl() const -> size_type {
```
Comment (BIGbadEL) [inline include/ftl/inspect_iterator.hpp:33]: noexcept?

File: include/ftl/inspect_iterator.hpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    return iterator_.count();
+  }
+
+  [[nodiscard]] auto enumerate_impl() const -> enumerate_iterator<inspect_iterator<Iter, Callable>> {
```
Comment (BIGbadEL) [inline include/ftl/inspect_iterator.hpp:37]: noexcept?

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

**Review by BIGbadEL — APPROVED** [review]

**Pull request merged**